EZP-25088: Search engine signal slots configuration per search engine

alongosz opened the pull request:
Related to [EZP-25088](https://jira.ez.no/browse/EZP-25088) and ~~#1660~~ #1707.

Goal: disable some signal slots when indexing for the Legacy Search Engine

**TODO**:
- [x] Create search signal slot configuration for each engine
- [x] Exclude some signal slots for LSE
- [x] Merge [Solr PR #51](https://github.com/ezsystems/ezplatform-solr-search-engine/pull/51) to run CI tests for Solr without errors
- [x] Change Signal Dispatcher to dispatch signals for relevant search engine only (configured per siteaccess)

alongosz commented:
Ready for review.
Ping @andrerom, @pspanja

pspanja commented on eZ/Bundle/EzPublishLegacySearchEngineBundle/EzPublishLegacySearchEngineBundle.php:
```diff
@@ -25,6 +27,9 @@ public function build(ContainerBuilder $container)
         $container->addCompilerPass(new SortClauseConverterPass());
+
+        $container->addCompilerPass(new FieldRegistryPass());
+        $container->addCompilerPass(new SignalSlotPass()); // Side note: to disable signal slots not needed for LSE, other search engine bundles need to be disabled in AppKernel
```
Don't suffix comments, put it on a separate line above.

pspanja commented:
On second thought, I think we can't limit having only one search engine activated in the app kernel.
The use-case is multi-site, multi-repository installations, where different sites should be able to use different engines.

The solution would be to use engine-specific tags and compiler passes, so we don't conflict on these.
WDYT @andrerom.

andrerom commented:
well depends, the PR here says only thing we want to differentiate is _some_ signals for legacy search engine? (And with some, is that _Location_ specific tags? Are they already separated by different tags?) Or are there further differences?

pspanja commented:
They use the same tags, but the real problem is they register to the same dispatcher. So I was wrong, separating the tags per-engine would not help -- we would still end up with slots from the different engine than configured for the siteaccess. We need to find a way to make slots vary per repository's search engine. I guess we can do that through repository factory, but I don't like the idea since it can't be compiled.

One other option is if/else per engine implementation in the slots.

alongosz commented:
Note: alternative approach to this PR: #1716.

andrerom commented:
Alternative approach:
If we instead introduce a factory that produces the dispatcher, this would leave dispatcher clean. As we in retrospective don't really care about site access changes mid way as they are used for switching between siteaccess on same repository, not different *(several things will break if you try that and expect things to work)*, hence there is no need to overcomplicate that part.

So for compiler pass that would mean something like:
```
diff --git a/eZ/Publish/Core/Base/Container/Compiler/Search/SignalSlotPass.php b/eZ/Publish/Core/Base/Container/Compiler/Search/SignalSlotPass.ph
index 79e4274..c679645 100644
--- a/eZ/Publish/Core/Base/Container/Compiler/Search/SignalSlotPass.php
+++ b/eZ/Publish/Core/Base/Container/Compiler/Search/SignalSlotPass.php
@@ -20,15 +20,29 @@ use Symfony\Component\DependencyInjection\Reference;
  */
 class SignalSlotPass implements CompilerPassInterface
 {
+    /**
+     * Name of a search engine for which Compiler Pass was instantiated.
+     *
+     * @var string
+     */
+    private $searchEngineName;
+
+    public function __construct($searchEngineName)
+    {
+        $this->searchEngineName = $searchEngineName;
+    }
+
     public function process(ContainerBuilder $container)
     {
         if (!$container->hasDefinition('ezpublish.signalslot.signal_dispatcher')) {
             return;
         }
 
-        $signalDispatcherDef = $container->getDefinition('ezpublish.signalslot.signal_dispatcher');
+        $signalDispatcherDef = $container->getDefinition('ezpublish.signalslot.signal_factory');
+        $tags = $container->findTaggedServiceIds('ezpublish.search.slot')
+            + $container->findTaggedServiceIds('ezpublish.search.slot.' . $this->searchEngineName);
 
-        foreach ($container->findTaggedServiceIds('ezpublish.search.slot') as $id => $attributes) {
+        foreach ($tags as $id => $attributes) {
             foreach ($attributes as $attribute) {
                 if (!isset($attribute['signal'])) {
                     throw new LogicException(
@@ -37,9 +51,11 @@ class SignalSlotPass implements CompilerPassInterface
                     );
                 }
 
+                // TODO: If possible we should build up array and attach all at once to reduce method calls
                 $signalDispatcherDef->addMethodCall(
                     'attach',
                     array(
+                        $this->searchEngineName,
                         $attribute['signal'],
                         new Reference($id),
                     )
```

andrerom commented on eZ/Publish/Core/Base/Container/Compiler/Search/SignalSlotPass.php:
```diff
@@ -20,31 +19,47 @@
+     */
+    private $searchEngineAlias;
+
+    public function __construct($searchEngineAlias = null)
```
why is this allowed to be null?

alongosz replied:
@andrerom: I wanted this to be a soft change, to avoid breaking Solr: https://github.com/ezsystems/ezplatform-solr-search-engine/blob/master/bundle/EzSystemsEzPlatformSolrSearchEngineBundle.php#L36

andrerom replied:
Hmm, then it's better that we introduce a new class and deprecate this one _(otherwise leave it as is)_, because the logic further down expects it to be set _(and it should)_.

andrerom commented on eZ/Publish/Core/Base/Container/Compiler/Search/SignalSlotPass.php:
```diff
@@ -20,31 +19,47 @@
                     throw new LogicException(
                         "Could not find 'signal' attribute on '$id' service, " .
-                        "which is mandatory for services tagged as 'ezpublish.persistence.solr.slot'"
+                        "which is mandatory for services tagged as '$searchEngineSlotTagName'"
```
`'ezpublish.search.slot' or the specific tags for given search engine.`

andrerom commented on eZ/Publish/Core/Base/Container/Compiler/Search/SignalSlotPass.php:
```diff
@@ -20,31 +19,47 @@
             }
         }
+
+        $signalDispatcherDef->addMethodCall('attachAll', $dispatcherCallsArguments);
```
good, but could we inject in constructor instead somehow _(makes the factory cleaner)_? /cc @pspanja

andrerom commented on eZ/Publish/Core/settings/tests/integration_legacy_core.yml:
```diff
@@ -11,3 +11,10 @@ services:
+        arguments:
+            - "%ezpublish.signalslot.signal_dispatcher.class%"
+            - "default"
+            - "%ezpublish.repositories%"
```
new line

alongosz commented on eZ/Publish/Core/settings/search_engines/elasticsearch/slots.yml:
```diff
@@ -0,0 +1,130 @@
+        parent: ezpublish.search.elasticsearch.slot
+        class: %ezpublish.search.elasticsearch.slot.publish_version.class%
+        tags:
+            - {name: ezpublish.search.elasticsearch.slot, signal: ContentService\PublishVersionSignal}
```
@andrerom: to be able to pass all signal slots to `SignalDispatcher` constructor in its factory, we should use full namespace for signals here. Should I go with this or attach namespace in `SignalDispatcherFactory` (as done in [`DefaultSignalDispatcher::attach`](https://github.com/alongosz/ezpublish-kernel/blob/88697520fae1de013d4fa50e0fd51c4daf43d8a4/eZ/Publish/Core/SignalSlot/SignalDispatcher/DefaultSignalDispatcher.php#L88-L92))?

andrerom replied:
@pspanja what do you think? BC wise we'll need to support old form for `ezpublish.search.slot`, so given that we might as well just keep the old form around for both inside the factory.

pspanja replied:
Agreed, it seems we have to do that for now.

alongosz replied:
Ok, changed in 2a9add2.

alongosz commented on eZ/Publish/Core/Base/Container/Compiler/Search/SearchEngineSignalSlotPass.php:
```diff
@@ -0,0 +1,65 @@
+                $signalDispatcherSlots[] = [
+                    'searchEngineAlias' => $this->searchEngineAlias,
+                    'signalIdentifier' => $attribute['signal'],
+                    'slot' => new Reference($id),
```
@andrerom: I managed to reduce number of calls adding slots to dispatcher, but needed to resign from ValueObject here, because [Symfony Container Builder resolves nested reference only for simple arrays](https://github.com/symfony/symfony/blob/2.8/src/Symfony/Component/DependencyInjection/ContainerBuilder.php#L983-L986). Not sure if you're ok with this solution :)

andrerom replied:
I wasn't really so fan of the Value object for this anyway so I'm more than fine with dropping it _(we don't use value objects for modeling service arguments)_ :)

Side: please don't rebase while PR is WIP, hard to see what change to refer to if it is no longer part of PR :)

andrerom replied:
But maybe better to model this per search engine to not have such a complex hash structure as argument.
So something like for instance:
```
function addSlotsForSearchEngine( $searchEngineAlias,  array( '<signal>' => array(), '<signal2>' => array(), ... ));
```

alongosz replied:
> But maybe better to model this per search engine to not have such a complex hash structure as argument.
> So something like for instance:
> ```
> function addSlotsForSearchEngine( $searchEngineAlias,  array( '<signal>' => array(), > '<signal2>' => array(), ... ));
> ```

Changed in fa296f5. It [simplifies this method](https://github.com/alongosz/ezpublish-kernel/blob/fa296f5ef2ad022c562792b1e77f2fa9e32938d8/eZ/Publish/Core/SignalSlot/SignalDispatcher/SignalDispatcherFactory.php#L59-L74) a lot. Thx @andrerom :)

andrerom commented on eZ/Publish/Core/Base/Container/Compiler/Search/SearchEngineSignalSlotPass.php:
```diff
@@ -0,0 +1,65 @@
+            return;
+        }
+
+        $signalDispatcherDef = $container->getDefinition('ezpublish.signalslot.signal_dispatcher.factory');
```
signalDispatcherFactoryDef

andrerom commented on eZ/Publish/Core/SignalSlot/SignalDispatcher/SignalDispatcherFactory.php:
```diff
@@ -0,0 +1,88 @@
+            $repositoryAlias = array_shift($aliases);
+        }
+        $this->repositorySettings = isset($repositoriesSettings[$repositoryAlias]) ? $repositoriesSettings[$repositoryAlias] : [];
+        $this->signalSlotMap = [];
```
can just as well set property to `[]` by default and skip this line right?

andrerom commented on eZ/Publish/Core/settings/containerBuilder.php:
```diff
@@ -47,12 +47,29 @@
 
 $containerBuilder->setParameter('ezpublish.kernel.root_dir', $installDir);
 
+// needed by @ezpublish.signalslot.signal_dispatcher:
```
ezpublish.signalslot.signal_dispatcher.factory

andrerom commented on eZ/Publish/SPI/Tests/FieldType/FileBaseIntegrationTest.php:
```diff
@@ -156,6 +156,24 @@ protected function getContainer()
             self::$tmpDir . '/var/ezdemo_site/storage'
         );
 
+        // needed by @ezpublish.signalslot.signal_dispatcher:
```
ezpublish.signalslot.signal_dispatcher.factory

andrerom commented on eZ/Publish/SPI/Tests/FieldType/BaseIntegrationTest.php:
```diff
@@ -591,6 +591,24 @@ protected function getContainer()
             $this->getDsn()
         );
 
+        // needed by @ezpublish.signalslot.signal_dispatcher:
```
ezpublish.signalslot.signal_dispatcher.factory

andrerom commented:
Ok, so remaining concern is if someone has search engine that still uses `SignalSlotPass`, this will lead to issues now _(will afaik attach additional signals on dispatcher)_.

@pspanja suggestions? Throw exception on use of it?

pspanja commented:
@andrerom 

> Or maybe it's no problem, until now people have had to only enable one engine at a time, so if they use old engine they still need to do that.

It was possible to use multiple engines before. But I think this should be OK anyway, since behavior remains the same; before this the slot would be active for all engines, and the same will be the case now.

pspanja commented on eZ/Publish/Core/SignalSlot/SignalDispatcher/SignalDispatcherFactory.php:
```diff
@@ -0,0 +1,85 @@
+ * @copyright Copyright (C) eZ Systems AS. All rights reserved.
+ * @license For full copyright and license information view LICENSE file distributed with this source code.
+ */
+namespace eZ\Publish\Core\SignalSlot\SignalDispatcher;
```
This should be rather placed in `eZ\Publish\Core\Base\Container\ApiLoader`, maybe in a separate dir `SignalSlot`.

pspanja commented on eZ/Publish/Core/settings/repository.yml:
```diff
@@ -149,4 +150,11 @@ services:
+        class: %ezpublish.signalslot.signal_dispatcher.factory.class%
+        arguments:
+            - "%ezpublish.signalslot.signal_dispatcher.class%"
+            - "$repository$"
```
This should not be here, since we don't know about `$repository$` or `%ezpublish.repositories%` at this level.

I propose you implement a `Core` version of `SignalDispatcherFactory`, which would receive search engine alias directly. Then you can extend that in the `EzPublishCoreBundle` and add knowledge about `$repository$` and `%ezpublish.repositories%`.
Maybe implement internal method `getSearchEngineAlias()`, in `Core` is can simply return injected alias, and in the bundle it can resolve the same from injected repository alias and repository settings (you will also have to override this service definition in the bundle).

Then you can also lose emulation of the full app by adding `ezpublish.repositories` param to the container builder.

andrerom replied:
@alongosz There are examples of different factories for integration tests vs app, for instnace reposiotry. Grab @pspanja on skype if you need some pointers ;)

Side: For later we should try to get rid of this separation to some extent as it causes issues whenever they are out of sync, and is causing us to have duplicated code for several things _(factories and setup of compiler passes)_.

alongosz commented:
> This should not be here, since we don't know about `$repository$` or `%ezpublish.repositories%` at this level.

Ok. In 3551642 I moved it to [eZ/Bundle/EzPublishCoreBundle/Resources/config/services.yml](https://github.com/alongosz/ezpublish-kernel/blob/3551642dd5267770d87b0c2a3faa3436b8d1f766/eZ/Bundle/EzPublishCoreBundle/Resources/config/services.yml#L213-L218)

> Maybe implement internal method `getSearchEngineAlias()`, in `Core` is can simply return injected alias, and in the bundle it can resolve the same from injected repository alias and repository settings

I think it is not needed due to the way [I implemented `SignalDispatcherFactory` in `Core`](https://github.com/alongosz/ezpublish-kernel/blob/3551642dd5267770d87b0c2a3faa3436b8d1f766/eZ/Publish/Core/Base/Container/ApiLoader/SignalSlot/SignalDispatcherFactory.php#L34-L62)

> (you will also have to override this service definition in the bundle).

The main problem with this is that [`eZ/Bundle/EzPublishCoreBundle/Resources/config/services.yml`](https://github.com/alongosz/ezpublish-kernel/blob/3551642dd5267770d87b0c2a3faa3436b8d1f766/eZ/Bundle/EzPublishCoreBundle/Resources/config/services.yml) is overriden by [`eZ/Publish/Core/settings/repository.yml`](https://github.com/alongosz/ezpublish-kernel/blob/3551642dd5267770d87b0c2a3faa3436b8d1f766/eZ/Publish/Core/settings/repository.yml) in [`EzPublishCoreExtension`](https://github.com/alongosz/ezpublish-kernel/blob/3551642dd5267770d87b0c2a3faa3436b8d1f766/eZ/Bundle/EzPublishCoreBundle/DependencyInjection/EzPublishCoreExtension.php#L305).

My solution to this is to remove `ezpublish.signalslot.signal_dispatcher.factory` from `repository.yml` and place it in [`eZ/Publish/Core/settings/tests/integration_legacy_core.yml`](https://github.com/alongosz/ezpublish-kernel/blob/3551642dd5267770d87b0c2a3faa3436b8d1f766/eZ/Publish/Core/settings/tests/integration_legacy_core.yml#L17) which will allow `phpunit.xml`tests to pass. Also [`eZ/Publish/Core/settings/tests/integration_legacy.yml`](https://github.com/alongosz/ezpublish-kernel/blob/3551642dd5267770d87b0c2a3faa3436b8d1f766/eZ/Publish/Core/settings/tests/integration_legacy.yml) and [`eZ/Publish/Core/settings/tests/integration_legacy_elasticsearch.yml`](https://github.com/alongosz/ezpublish-kernel/blob/3551642dd5267770d87b0c2a3faa3436b8d1f766/eZ/Publish/Core/settings/tests/integration_legacy_elasticsearch.yml) have now proper, context-related, overrides.

I tested this using ezplatform web app with repositories using Legacy and Elasticsearch engines and it seems to work.

However this breaks Solr which needs to be updated. Updated Solr would require Kernel from this PR which leads to circular dependency. I'll try to create a PR for Solr introducing "soft" change for this to work (not sure if possible).

ping @pspanja, @andrerom.

pspanja commented on eZ/Bundle/EzPublishCoreBundle/Resources/config/services.yml:
```diff
@@ -198,3 +201,18 @@ services:
     ez_query:
         alias: ezpublish.controller.query
+
+    ezpublish.signalslot.signal_dispatcher_transaction_wrapper:
```
Why do you redefine this?

alongosz replied:
Removed in 0b1e1ba.

pspanja commented on eZ/Bundle/EzPublishCoreBundle/Resources/config/services.yml:
```diff
@@ -198,3 +201,18 @@ services:
+        class: %ezpublish.signalslot.signal_dispatcher_transaction_wrapper.class%
+        arguments: [@ezpublish.signalslot.signal_dispatcher]
+
+    ezpublish.signalslot.signal_dispatcher:
```
Why is this redefined?

alongosz replied:
Leftover from my tests, my mistake. Removed in 0b1e1ba.

pspanja commented on eZ/Bundle/EzPublishCoreBundle/Resources/config/services.yml:
```diff
@@ -198,3 +201,18 @@ services:
+        class: %ezpublish.signalslot.signal_dispatcher.class%
+        factory: ["@ezpublish.signalslot.signal_dispatcher.factory", buildSignalDispatcher]
+
+    ezpublish.signalslot.signal_dispatcher.factory:
```
You should rather place this in `eZ/Bundle/EzPublishCoreBundle/Resources/config/papi.yml` to achieve override.

alongosz replied:
Thanks, this solved all problems :)

pspanja commented:
> I think it is not needed due to the way I implemented SignalDispatcherFactory in Core

I would rather avoid doing the resolving in the constructor, it's not a big problem as the logic is light, but it would still be better to have a separate method to do it on demand (particularly if we need to change search engine later on for example).

alongosz commented on eZ/Bundle/EzPublishCoreBundle/ApiLoader/SignalSlot/SignalDispatcherFactory.php:
```diff
@@ -0,0 +1,34 @@
+            $aliases = array_keys($repositoriesSettings);
+            $repositoryAlias = array_shift($aliases);
+        }
+        $searchEngineAlias = isset($repositoriesSettings[$repositoryAlias]['search']['engine']) ? $repositoriesSettings[$repositoryAlias]['search']['engine'] : [];
```
> I would rather avoid doing the resolving in the constructor, it's not a big problem as the logic is light, but it would still be better to have a separate method to do it on demand (particularly if we need to change search engine later on for example).

@pspanja: should I move this whole logic (lines 27-31) to `getSearchEngineAlias` (including resolving repository settings) or just store here single repository settings and then in `getSearchEngineAlias` only return `$this->repositorySettings['search']['engine']`?

pspanja replied:
Move the whole logic, just store repository alias and settings to private properties here.

alongosz commented:
Ready for review after changes
ping @pspanja, @andrerom

andrerom commented:
+1 but @pspanja might have some more comments.

pspanja commented on eZ/Bundle/EzPublishCoreBundle/ApiLoader/SignalSlot/SignalDispatcherFactory.php:
```diff
@@ -0,0 +1,54 @@
+    ) {
+        $this->repositoryAlias = $repositoryAlias;
+        $this->repositoriesSettings = $repositoriesSettings;
+        parent::__construct($signalDispatcherClass, $this->getSearchEngineAlias());
```
There is no need to resolve the search engine alias here, so you can just pass `null` for the second argument.

alongosz replied:
Ok, fixed in 5a23d73

pspanja commented:
+1 aside from one remark.
